From 66bbfc4f571fb73733e8081a4e33b5a8601a4c3a Mon Sep 17 00:00:00 2001
From: "Guoniu.Zhou" <guoniu.zhou@nxp.com>
Date: Fri, 9 Nov 2018 18:37:09 +0800
Subject: [PATCH 5096/5242] MLK-20326-1: ISI: enable horizonal and vertical
 flip for camera

commit  9e33846dd4b77df35a8de25aa8477b5ca58ba228 from
https://source.codeaurora.org/external/imx/linux-imx.git

Enable horizonal and vertical flip for camera image

Signed-off-by: Guoniu.Zhou <guoniu.zhou@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/imx8/mxc-isi-cap.c |   12 ++++++++++--
 drivers/media/platform/imx8/mxc-isi-hw.c  |    2 +-
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/drivers/media/platform/imx8/mxc-isi-cap.c b/drivers/media/platform/imx8/mxc-isi-cap.c
index 664eeb5..4cb9444 100644
--- a/drivers/media/platform/imx8/mxc-isi-cap.c
+++ b/drivers/media/platform/imx8/mxc-isi-cap.c
@@ -547,16 +547,24 @@ static int mxc_isi_s_ctrl(struct v4l2_ctrl *ctrl)
 
 	switch (ctrl->id) {
 	case V4L2_CID_HFLIP:
-		mxc_isi->hflip = ctrl->val;
+		if (ctrl->val < 0)
+			return -EINVAL;
+		mxc_isi->hflip = (ctrl->val > 0) ? 1 : 0;
 		break;
 
 	case V4L2_CID_VFLIP:
-		mxc_isi->vflip = ctrl->val;
+		if (ctrl->val < 0)
+			return -EINVAL;
+		mxc_isi->vflip = (ctrl->val > 0) ? 1 : 0;
 		break;
 
 	case V4L2_CID_ALPHA_COMPONENT:
 		mxc_isi->alpha = ctrl->val;
 		break;
+
+	default:
+		dev_err(&mxc_isi->pdev->dev, "%s: Not support %d CID\n", __func__, ctrl->id);
+		return -EINVAL;
 	}
 
 	spin_unlock_irqrestore(&mxc_isi->slock, flags);
diff --git a/drivers/media/platform/imx8/mxc-isi-hw.c b/drivers/media/platform/imx8/mxc-isi-hw.c
index 0e99dd1..c5bc711 100644
--- a/drivers/media/platform/imx8/mxc-isi-hw.c
+++ b/drivers/media/platform/imx8/mxc-isi-hw.c
@@ -571,10 +571,10 @@ void mxc_isi_channel_config(struct mxc_isi_dev *mxc_isi)
 	writel(val, mxc_isi->regs + CHNL_OUT_BUF_PITCH);
 
 	/* TODO */
+	mxc_isi_channel_set_flip(mxc_isi);
 #if 0
 	mxc_isi_channel_set_crop(mxc_isi);
 
-	mxc_isi_channel_set_flip(mxc_isi);
 	if (mxc_isi->alphaen)
 		mxc_isi_channel_set_alpha(mxc_isi);
 
-- 
1.7.9.5

